xcode/Subconscious/Shared/Components/Common/SubtextTextViewRepresentable.swift

```diff
@@ -164,38 +164,55 @@ struct SubtextTextViewRepresentable: UIViewRepresentable {
         }
     }
 
-    class SubtextTextParagraph: NSTextParagraph {
-        private(set) var subtext: Subtext
-
-        init(attributedString: NSAttributedString) {
-            SubtextTextViewRepresentable.logger.debug(
-                "SubtextTextParagraph: parse and render markup"
-            )
-            let subtext = Subtext(markup: attributedString.string)
-            let renderedAttributedString = subtext.renderAttributedString(
-                url: SubtextTextViewRepresentable.toSubURL
-            )
-            self.subtext = subtext
-            super.init(attributedString: renderedAttributedString)
-        }
-    }
-
     //  MARK: Coordinator
-    class Coordinator: NSObject, UITextViewDelegate, NSTextContentStorageDelegate, NSTextContentManagerDelegate, NSTextLayoutManagerDelegate {
+    class Coordinator: NSObject, UITextViewDelegate, NSTextContentStorageDelegate, NSTextStorageDelegate, NSTextContentManagerDelegate, NSTextLayoutManagerDelegate {
         /// Is event happening during updateUIView?
         /// Used to avoid setting properties in events during view updates, as
         /// that would cause feedback cycles where an update triggers an event,
         /// which triggers an update, which triggers an event, etc.
         var isUIViewUpdating: Bool
         var representable: SubtextTextViewRepresentable
         var renderTranscludeBlocks: Bool = false
+        var subtext: Subtext? = nil
 
         init(
             representable: SubtextTextViewRepresentable
         ) {
             self.isUIViewUpdating = false
             self.representable = representable
         }
+        
+        /// NSTextStorageDelegate method
+        /// Handle markup rendering, just before processEditing is fired.
+        /// It is important that we render markup in `willProcessEditing`
+        /// because it happens BEFORE font substitution. Rendering before font
+        /// substitution gives the OS a chance to replace fonts for things like
+        /// Emoji or Unicode characters when your font does not support them.
+        /// See:
+        /// https://github.com/gordonbrander/subconscious/wiki/nstextstorage-font-substitution-and-missing-text
+        ///
+        /// 2022-03-17 Gordon Brander
+        func textStorage(
+          _ textStorage: NSTextStorage,
+          willProcessEditing: NSTextStorage.EditActions,
+          range: NSRange,
+          changeInLength: Int
+        ) {
+            SubtextTextViewRepresentable.logger.debug(
+              "textStorage: render markup attributes"
+            )
+            textStorage.setAttributes(
```

Comment:
If I understand correctly, this is clearing all attributes from string.

I think we should always clear all attributes from string before rendering new attributes with `renderAttributesOf`.

Let's move this to `Subtext.renderAttributesOf`.

Reply:
This is the way it _used_ to work before we moved to TK2. Regardless, I've moved this into `renderAttributesOf` now.